### ISSUE TITLE
Version converter 11 -> 12 fails with "IndexError: Input s is undefined!"

### ISSUE THREAD

**Issue by matteosal** [onnx/onnx#2873]
# Bug Report

### Describe the bug
The version converter fails on the attached model (and not becase the adapter is missing)

### System information
- OS Platform and Distribution: Ubuntu 18.04  
- ONNX version: 1.7 release commit
- Python version: 3.6.9
- GCC/Compiler version: 7.5.0
- CMake version: 3.17.3
- Protobuf version: 3.12

### Reproduction instructions
```
import onnx
from onnx import version_converter, helper

# Preprocessing: load the model to be converted.
model_path = "path/to/shared-bn-arrays-v11.onnx"
original_model = onnx.load(model_path)

converted_model = version_converter.convert_version(original_model, 12)
print("done!")
```
The script doesn't get to the final print, but fails with 
```
Traceback (most recent call last):
  File "script.py", line 8, in <module>
    converted_model = version_converter.convert_version(original_model, 12)
  File "/usr/local/lib/python3.6/dist-packages/onnx-1.7.0-py3.6-linux-x86_64.egg/onnx/version_converter.py", line 166, in convert_version
    converted_model_str = C.convert_version(model_str, target_version)
IndexError: Input s is undefined!
```

- ONNX model
[shared-bn-arrays-v11.zip](https://github.com/onnx/onnx/files/4858440/shared-bn-arrays-v11.zip)

### Expected behavior
The script should not fail and properly upgrade the model (or at least fail saying there are no adapters)

**Comment by askhade** [comment]
There is an underlying issue in version converter. It relies on the C++ IR which I believe has not been updated after IR v3. Because of this I think it expects initializers also be added as graph inputs. 
If you try to change the version of your model to IRv3 or create a model with initializers also as inputs then I think this will work.

**Comment by matteosal** [comment]
Yes, by including all the initializers in the graph inputs it works, thanks. So this seems related to [this other issue](https://github.com/onnx/onnx/issues/2874) I created today.

**Comment by jcwchen** [comment]
Fortunately, it can be fixed by correcting `shape_inference` about the IR issue because `version_converter` uses `shape_inference` before the conversion. This PR https://github.com/onnx/onnx/pull/2901 should help.

**Comment by jcwchen** [comment]
It should be resolved by https://github.com/onnx/onnx/pull/2901. Please reopen it if you still encounter this issue. Thanks.

**Comment by jcwchen** [comment]
For fix shape_inference, we eventually decided to not update the input from initializer to the graph so version_converter will still encounter this issue... Reopen it until it has been fixed. Sorry for the confusion.

**Comment by ahyangyi** [comment]
@jcwchen 
There is a special case: converting 10->10 still triggers this bug.

Fixing this special case, however, might be much easier than other cases, since the converter can and should shortcut and stop early.

**Comment by matteosal** [comment]
@jcwchen Any plans to fix this? 
I have started adding more adapters for the version converter (will open a PR when ready), but this bug makes such endeavor less useful (because conversion might fail anyway regardless of what adapters are present).
It also limits the ability to test conversion on real-world models. All these detection models from the zoo are affected by this bug:
* https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/ssd-mobilenetv1
* https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/retinanet
* https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/tiny-yolov3
* https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/yolov2-coco
* https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/yolov3
* https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/yolov4

**Comment by jcwchen** [comment]
Hi @matteosal,
I agree that it is an essential issue for version_converter because I did see some users complained about it. I tried to fix it once but I could not find a proper way and currently I don't more bandwidth to deal with this... I will tag it as an enhancement and post it here https://github.com/onnx/onnx/issues/3225 to wait someone who is interested to pick it up. Thank you.

**Comment by mrsabhar** [comment]
I'm facing same issue with ONNX 1.9. 
Trying to convert model from Opset 9 to Opset 12 but getting error "Input 1 is undefined". 

OS : Windows 10
Python : Anaconda 3.7.10
ONNX version - 1.9

model_path = sys.argv[1]
original_model = onnx.load(model_path)
print(original_model.graph.input)
print('The model before conversion:\n{}'.format(original_model))
onnx.checker.check_model(original_model)

converted_model = version_converter.convert_version(original_model, 12)

Error - 
opset_import {
  version: 9
}

Traceback (most recent call last):
  File "convert_opset.py", line 11, in <module>
    converted_model = version_converter.convert_version(original_model, 12)
  File "C:\Users\xyz\Anaconda3\envs\opset\lib\site-packages\onnx\version_converter.py", line 174, in convert_version
    converted_model_str = C.convert_version(model_str, target_version)
IndexError: Input 1 is undefined!

I did try with ONNX 1.8 but still same result. 

Input to model is not dynamic - Input [1,3,24,224]. Above script works with ResNet50 but fails with specific models.

**Comment by jcwchen** [comment]
Hi @mrsabhar,
It hasn't been fixed yet, but there is a workaround for it, add initializer into input: https://github.com/onnx/onnx/issues/2660#issuecomment-605874784 You can try the code here and see whether it resolves your issue. Thanks

**Comment by Haydnspass** [comment]
@jcwchen Thanks for pointing it out. Since I am kind of new to ONNX could you point me to where and how I need to apply the workaround? My small script is:

```python
import onnx
from onnx import version_converter, helper


def add_value_info_for_constants(model : onnx.ModelProto):
   # ...
   return


model = onnx.load('[model path]')
model = onnx.shape_inference.infer_shapes(model)
converted_model = version_converter.convert_version(model, 9)

onnx.save(converted_model, '[new_path]')
```

**Comment by jcwchen** [comment]
Hi @Haydnspass,
Just use the function "add_value_info_for_constants" from [here](https://github.com/onnx/onnx/issues/2660#issuecomment-605874784) before the versioon_converter.
```
model = onnx.shape_inference.infer_shapes(model)
add_value_info_for_constants(model)
converted_model = version_converter.convert_version(model, 9)
```

**Comment by zhuguiqian** [comment]
> Hi @Haydnspass,
> Just use the function "add_value_info_for_constants" from [here](https://github.com/onnx/onnx/issues/2660#issuecomment-605874784) before the versioon_converter.
> 
> ```
> model = onnx.shape_inference.infer_shapes(model)
> add_value_info_for_constants(model)
> converted_model = version_converter.convert_version(model, 9)
> ```

hi, @jcwchen
I find  function "add_value_info_for_constants"  require ir_version>=4, but my model's ir_version is 3, I tried the mentioned function, it didn't work, also with undefined error.
Any other workaround for ir_version 3?
![image](https://user-images.githubusercontent.com/7693110/123600393-38045980-d829-11eb-94a5-978f853ee88d.png)

**Comment by zhuguiqian** [comment]
> > hi, @jcwchen
> > I find function "add_value_info_for_constants" require ir_version>=4, but my model's ir_version is 3, I tried the mentioned function, it didn't work, also with undefined error.
> > Any other workaround for ir_version 3?
> > ![image](https://user-images.githubusercontent.com/7693110/123600393-38045980-d829-11eb-94a5-978f853ee88d.png)
> 
> Hi @zhuguiqian,
> 
> Actually this workaround only applies to model whose IR_VERSION >= 4 because that bug does not exist in older models (IR_VERSION < 4). Old IR version needs graph.initializer to be included in graph.input as well. If your model is IR_VERSION=3 and it encounters an undefined input error, it should be another issue instead of this one.

OK, thanks

**Comment by mikeseven** [comment]
still not working, even with IR 6 and latest code on master today.

**Comment by mrsabhar-Intel** [comment]
I tried with workaround too but still not working. Conversion fails from opset10 --> Opset13 using ONNX 1.10.1. 

converted_model_str = C.convert_version(model_str, target_version)
IndexError: Input conv1.conv2d.weight is undefined!

**Comment by LingCoderSonoscape** [comment]
Hi I tried to convert my model from opset-version 11 to opset-version 9, but it threw an IndexError which is like below:
**_IndexError: Input 633 is undefined!_**
The onnx model I got was exported from mmsegmentation-framework using pytorch-1.7. 
Any update so far that can handle this problem?

**Comment by LingCoderSonoscape** [comment]
Hi I tried to convert my model from opset-version 11 to opset-version 9, but it threw an IndexError which is like below:
**_IndexError: Input 633 is undefined!_**
The onnx model I got was exported from mmsegmentation-framework using pytorch-1.7. 
Any update so far that can handle this problem?

**Comment by LingCoderSonoscape** [comment]
@jcwchen Hi! thank you for your update, I have tried the new version of the package and it did work on creating the graph by using the _**add_value_info_for_constants()**_ function. However, when I ran the _**version_converter.convert_verison**_ using the new model. It threw out the segmentation fault. 
The opset-version I tried to convert is from 11 to 9. Is this converting function still in progress or I did not use it properly?
Looking forward to your reply

**Comment by jcwchen** [comment]
Hi @LingCoderSonoscape,
If you did see the initializers have been added into graph input after applying `add_value_info_for_constants` and the version converter is still not working, it is probably caused by other issues instead of the one this thread is talking about. If that is the case, I would suggest you open a new issue and please also provide the model for me to repro. Thank you!


> I have tried the new version of the package

You mean ONNX 1.10.1, right?

**Comment by mrsabhar-Intel** [comment]
That is correct. ONNX 1.10.1 is not working for me.

**Comment by jcwchen** [comment]
Please note that
1. The fix (consider initializer as node input due to IR gap) has not been merged yet (so it's not in any official release)
2. If you bump into this error, you can first try the workaround (use python script to manually add graph initializer into graph input) mentioned in this thread
3. If the workaround above does not work for your case, please raise a new issue in ONNX because that error is caused by other issues.

Anyone interested free feel to review this PR to fix the IR gap issue: https://github.com/onnx/onnx/pull/3676 Thank you everyone!

**Comment by ZhangXiaoying0116** [comment]
@

> Please note that
> 
> 1. The fix (consider initializer as node input due to IR gap) has not been merged yet (so it's not in any official release)
> 2. If you bump into this error, you can first try the workaround (use python script to manually add graph initializer into graph input) mentioned in this thread
> 3. If the workaround above does not work for your case, please raise a new issue in ONNX because that error is caused by other issues.
> 
> Anyone interested free feel to review this PR to fix the IR gap issue: #3676 Thank you everyone!

python script add_value_info_for_constants won't work because this func just adds init to value_info instead of input. I think graph.value_info.add() should be replaced with graph.input.add() .

**Comment by jcwchen** [comment]
Hi @ZhangXiaoying0116,
Thank you for catching this! I referenced a wrong code... This one should be the right one: https://github.com/onnx/onnx/issues/2995#issuecomment-687631931 (add_input_from_initializer). I will correct my response above as well.
